packages/ember-glimmer/tests/integration/helpers/custom-helper-test.js

```diff
@@ -22,4 +26,431 @@ moduleFor('Helpers test: custom helpers', class extends RenderingTest {
+    assert.strictEqual(params[1], 'rich', 'second argument is rich');
+    assert.strictEqual(hash.last, 'sam', 'hash.last argument is sam');
+  }
+
+  ['@test helper receives params, hash']() {
+    let params, hash;
+
+    this.registerHelper('hello-world', {
+      compute(_params, _hash) {
+        params = _params;
+        hash = _hash;
+      }
+    });
+
+    this.render('{{hello-world name "rich" last="sam"}}', {
+      name: 'bob'
+    });
+
+    assert.strictEqual(params[0], 'bob', 'first argument is bob');
+    assert.strictEqual(params[1], 'rich', 'second argument is rich');
+    assert.strictEqual(hash.last, 'sam', 'hash.last argument is sam');
+
+    this.runTask(() => this.rerender());
+
+    assert.strictEqual(params[0], 'bob', 'first argument is bob');
+    assert.strictEqual(params[1], 'rich', 'second argument is rich');
+    assert.strictEqual(hash.last, 'sam', 'hash.last argument is sam');
+
+    this.runTask(() => set(this.context, 'name', 'sal'));
+
+    assert.strictEqual(params[0], 'sal', 'first argument is sal');
+    assert.strictEqual(params[1], 'rich', 'second argument is rich');
+    assert.strictEqual(hash.last, 'sam', 'hash.last argument is sam');
+
+    this.runTask(() => set(this.context, 'name', 'bob'));
+
+    assert.strictEqual(params[0], 'bob', 'first argument is bob');
+    assert.strictEqual(params[1], 'rich', 'second argument is rich');
+    assert.strictEqual(hash.last, 'sam', 'hash.last argument is sam');
+  }
+
+  ['@test helper usable in subexpressions']() {
+    this.registerHelper('join-words', {
+      compute(params) {
+        return params.join(' ');
+      }
+    });
+
+    this.render(
+      `{{join-words "Who"
+                   (join-words "overcomes" "by")
+                   "force"
+                   (join-words (join-words "hath overcome but" "half"))
+                   (join-words "his" (join-words "foe"))}}`);
+
+    this.assertText('Who overcomes by force hath overcome but half his foe');
+
+    this.runTask(() => this.rerender());
+
+    this.assertText('Who overcomes by force hath overcome but half his foe');
+  }
+
+  ['@htmlbars shorthand helper not usable with a block']() {
+    this.registerHelper('some-helper', () => {});
+
+    expectAssertion(() => {
+      this.render(`{{#some-helper}}{{/some-helper}}`);
+    }, /Helpers may not be used in the block form/);
+  }
+
+  ['@htmlbars helper not usable with a block']() {
+    this.registerHelper('some-helper', {
+      compute() {
+      }
+    });
+
+    expectAssertion(() => {
+      this.render(`{{#some-helper}}{{/some-helper}}`);
+    }, /Helpers may not be used in the block form/);
+  }
+
+  ['@htmlbars shorthand helper not usable within element']() {
+    this.registerHelper('some-helper', () => {});
+
+    expectAssertion(() => {
+      this.render(`<div {{some-helper}}></div>`);
+    }, /Helpers may not be used in the element form/);
+  }
+
+  ['@htmlbars helper not usable within element']() {
```

Comment:
This is marked as `@htmlbars` because it does not generate an error in Glimmer2, but it should.